msal/imds.py

```diff
@@ -0,0 +1,164 @@
+# Copyright (c) Microsoft Corporation.
+# All rights reserved.
+#
+# This code is licensed under the MIT License.
+import json
+import logging
+import os
+import socket
+import time
+try:  # Python 2
+    from urlparse import urlparse
+except:  # Python 3
+    from urllib.parse import urlparse
+
+logger = logging.getLogger(__name__)
+
+def _scope_to_resource(scope):  # This is an experimental reasonable-effort approach
+    u = urlparse(scope)
+    if u.scheme:
+        return "{}://{}".format(u.scheme, u.netloc)
+    return scope  # There is no much else we can do here
+
+
+def _obtain_token(http_client, resource, client_id=None):
+    if "IDENTITY_ENDPOINT" in os.environ and "IDENTITY_HEADER" in os.environ:
+        return _obtain_token_on_app_service(
+            http_client, os.environ["IDENTITY_ENDPOINT"], os.environ["IDENTITY_HEADER"],
+            resource, client_id=client_id)
+    return _obtain_token_on_azure_vm(http_client, resource, client_id=client_id)
+
+
+def _obtain_token_on_azure_vm(http_client, resource, client_id=None):
```

Comment:
Besides `client_id`, `object_id` and `msi_res_id` should also be supported:

- https://github.com/Azure/msrestazure-for-python/blob/8e3def8c9927a349596e575de93df76de7a3a2b6/msrestazure/azure_active_directory.py#L468
- https://github.com/MicrosoftDocs/azure-docs/issues/15288#issuecomment-619531505
- https://learn.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http